### PR TITLE
Add support for PostgreSQL databases

### LINKED ISSUE #177: Add support for creating PostgreSQL databases

### PR DESCRIPTION
**Pull request by h3llix**
Closes #177

### CODE REVIEW

File: docs/content/configurations/docker-images.md [hunk 1]
```diff
@@ -17,8 +17,9 @@ golang = "sdsws/golang:1.1"
 ruby = "sdsws/ruby:1.0"
 mysql = "mysql:5.7"
 mongodb = "mongo:4.2.1"
+postgres ="postgres:latest"
```
Comment (supra08) [inline docs/content/configurations/docker-images.md]: `postgresql`

File: config.sample.toml [hunk 2]
```diff
@@ -182,10 +184,20 @@ container_port = 33061  # Port on which the MySQL server container will run
...
+[services.kaen.postgresql.env]
+POSTGRES_USER = "YOUR_ROOT_NAME"   # Root user of PostgreSQL server inside the container
+POSTGRES_PASSWORD = "YOUR_ROOT_PASSWORD"   # Root password of PostgreSQL server inside the container
+
+# Configuration for PostgreSQL database server managed by `Kaen`
 [services.kaen.mongodb]
-plugin = false  # Deploy MongoDB server and let `Kaen` manage it?
-container_port = 27018  # Port on which the MongoDB server container will run
+plugin = false  # Deploy PostgreSQL server and let `Kaen` manage it
+container_port = 27018  # Port on which the PostgreSQL server container will run
```
Comment (supra08) [inline config.sample.toml]: MongoDB

File: config.sample.toml [hunk 3]
```diff
@@ -182,10 +184,20 @@ container_port = 33061  # Port on which the MySQL server container will run
...
+[services.kaen.postgresql]
+plugin = false  # Deploy PostgreSQL server and let `Kaen` manage it?
+container_port = 29121  # Port on which the PostgreSQL server container will run
+
+# Environment variables for PostgreSQL docker container.
+[services.kaen.postgresql.env]
+POSTGRES_USER = "YOUR_ROOT_NAME"   # Root user of PostgreSQL server inside the container
+POSTGRES_PASSWORD = "YOUR_ROOT_PASSWORD"   # Root password of PostgreSQL server inside the container
+
+# Configuration for PostgreSQL database server managed by `Kaen`
```
Comment (supra08) [inline config.sample.toml]: MongoDB

**Review by supra08 — CHANGES_REQUESTED** [review]

**Review by supra08 — APPROVED** [review]

File: .gitignore [hunk 4]
```diff
@@ -5,10 +5,13 @@ config.toml
...
+postgresql-storage/
 __debug_bin
 bin/
 .vscode/
 site/
 .DS_Store
 coverage.txt
 releases/
+
+
```
Comment (alphadose) [inline .gitignore]: remove the last extra line

File: go.mod [hunk 5]
```diff
@@ -20,12 +20,14 @@ require (
...
 	github.com/go-redis/redis v6.15.6+incompatible
 	github.com/go-sql-driver/mysql v1.4.1
-	github.com/go-stack/stack v1.8.0 // indirect
 	github.com/golang/protobuf v1.3.2
 	github.com/golang/snappy v0.0.1 // indirect
 	github.com/google/go-cmp v0.3.1 // indirect
 	github.com/gorilla/websocket v1.4.1 // indirect
 	github.com/hashicorp/go-multierror v1.0.0 // indirect
+	github.com/jackc/fake v0.0.0-20150926172116-812a484cc733 // indirect
+	github.com/jackc/pgx v3.6.2+incompatible
```
Comment (alphadose) [inline go.mod]: this version is not required.

Remove this and run `go mod vendor`

**Review by alphadose — CHANGES_REQUESTED** [review]

File: docs/content/configurations/docker-images.md [hunk 6]
```diff
@@ -17,8 +17,9 @@ golang = "sdsws/golang:1.1"
 ruby = "sdsws/ruby:1.0"
 mysql = "mysql:5.7"
 mongodb = "mongo:4.2.1"
+postgresql ="postgres:latest"
```
Comment (alphadose) [inline docs/content/configurations/docker-images.md]: Change this to `postgresql = "postgres:12.2-alpine"`

**Review by alphadose — CHANGES_REQUESTED** [review]

File: docs/content/configurations/docker-images.md (hunk 6 above)
Comment (h3llix) [reply docs/content/configurations/docker-images.md]: Fixed . made changes to overview also

**Comment by supra08** [comment]
Add the postgre image in the `overview.md`. Rest LGTM, we can merge it after that.

File: .gitignore [hunk 7]
```diff
@@ -5,10 +5,11 @@ config.toml
...
 mongodb-storage/
+postgresql-storage/
 __debug_bin
 bin/
 .vscode/
 site/
 .DS_Store
 coverage.txt
-releases/
+releases/
```
Comment (alphadose) [inline .gitignore]: add new line

File: lib/database/postgresql.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+func CreatePostgresqlDB(db types.Database) error {
+	ctx := context.Background()
+	connection := fmt.Sprintf("postgres://%v:%v@localhost:%d/%v", postgresqlRootUser, postgresqlPassword, postgresqlPort, postgresqlRootUser)
+	conn, err := pgx.Connect(ctx, connection)
+	if err != nil {
+		return fmt.Errorf("Error while creating the database : %s", err)
+	}
+	defer conn.Close(ctx)
+
+	if _, err = conn.Exec(ctx, "CREATE DATABASE "+db.GetName()); err != nil {
```
Comment (alphadose) [inline lib/database/postgresql.go]: use `fmt.Sprintf`

File: lib/database/postgresql.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	ctx := context.Background()
+
+	connection := fmt.Sprintf("postgres://%v:%v@localhost:%d/%v", postgresqlRootUser, postgresqlPassword, postgresqlPort, postgresqlRootUser)
+	conn, err := pgx.Connect(ctx, connection)
+	if err != nil {
+		return fmt.Errorf("Error while creating the database : %s", err)
+	}
+	defer conn.Close(ctx)
+
+	if _, err = conn.Exec(ctx, "DROP DATABASE IF EXISTS "+databaseName); err != nil {
```
Comment (alphadose) [inline lib/database/postgresql.go]: use `fmt.Sprintf`

**Review by alphadose — CHANGES_REQUESTED** [review]

**Review by alphadose — APPROVED** [review]

File: lib/database/postgresql.go [hunk 10]
```diff
@@ -42,6 +42,21 @@ func CreatePostgresqlDB(db types.Database) error {
 	if _, err = conn.Exec(ctx, query); err != nil {
 		return fmt.Errorf("Error while creating the database : %s", err)
 	}
+
+	query = fmt.Sprintf("REVOKE CONNECT ON DATABASE %s FROM PUBLIC;", db.GetName())
+	if _, err = conn.Exec(ctx, query); err != nil {
+		return fmt.Errorf("Error while revoking permissions : %s", err)
+	}
+
+	query = fmt.Sprintf("REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA pg_catalog  FROM PUBLIC;")
```
Comment (supra08) [inline lib/database/postgresql.go]: remove the extra space after `pg_catalog`

File: lib/database/postgresql.go [hunk 11]
```diff
@@ -42,6 +42,21 @@ func CreatePostgresqlDB(db types.Database) error {
...
+	if _, err = conn.Exec(ctx, query); err != nil {
+		return fmt.Errorf("Error while revoking permissions : %s", err)
+	}
+
+	query = fmt.Sprintf("REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA information_schema FROM PUBLIC;")
```
Comment (supra08) [inline lib/database/postgresql.go]: @alphadose I'm not sure about this but is revoking permissions from just `information_schema` a good idea? Wouldn't there be any action to revoke all access from the user after the db creation, similar to mysql?

Comment (supra08) [reply lib/database/postgresql.go]: Oh alright this has been resolved, good.

**Review by supra08 — APPROVED** [review]

**Pull request merged**